lib/spack/spack/spec.py

```diff
@@ -1825,6 +1829,74 @@ def process_hash_bit_prefix(self, bits):
         """Get the first <bits> bits of the DAG hash as an integer type."""
         return spack.util.hash.base32_prefix_bits(self.process_hash(), bits)
 
+    def _lookup_hash(self):
+        """Lookup just one spec with an abstract hash, returning a spec from the the environment,
+        store, or finally, binary caches."""
+        import spack.environment
+
+        matches = []
+        active_env = spack.environment.active_environment()
+
+        if active_env:
+            env_matches = active_env.get_by_hash(self.abstract_hash) or []
+            matches = [m for m in env_matches if m._satisfies(self)]
+        if not matches:
+            db_matches = spack.store.db.get_by_hash(self.abstract_hash) or []
+            matches = [m for m in db_matches if m._satisfies(self)]
+        if not matches:
+            query = spack.binary_distribution.BinaryCacheQuery(True)
+            remote_matches = query("/" + self.abstract_hash) or []
+            matches = [m for m in remote_matches if m._satisfies(self)]
+        if not matches:
+            raise InvalidHashError(self, self.abstract_hash)
+
+        if len(matches) != 1:
+            raise spack.spec.AmbiguousHashError(
+                f"Multiple packages specify hash beginning '{self.abstract_hash}'.", *matches
+            )
+
+        return matches[0]
+
+    def lookup_hash(self):
+        """Given a spec with an abstract hash, return a copy of the spec with all properties and
+        dependencies by looking up the hash in the environment, store, or finally, binary caches.
+        This is non-destructive."""
+        if self.concrete or not any(node.abstract_hash for node in self.traverse()):
+            return self
+
+        spec = self.copy(deps=False)
+        # root spec is replaced
+        if spec.abstract_hash:
+            new = self._lookup_hash()
+            spec._dup(new)
+            return spec
+
+        # Get dependencies that need to be replaced
+        for node in self.traverse(root=False):
+            if node.abstract_hash:
+                new = node._lookup_hash()
+                spec._add_dependency(new, deptypes=())
+
+        # reattach nodes that were not otherwise satisfied by new dependencies
+        for node in self.traverse(root=False):
+            if not any(n._satisfies(node) for n in spec.traverse()):
+                spec._add_dependency(node.copy(), deptypes=())
```

Comment:
I'm not 100% sure this is the right semantic, because it can return a contradiction.  Suppose you have:

```
foo ^/abc ^zlib@1.2
```

and the only lookup of `/abc` has `zlib@1.3` somewhere in the DAG.  This logic may give you a spec with a contradiction in it.  It *could* concretize if `/abc` is a build dependency or if it's `zlib` only reachable through build dependencies, since we allow multiple versions of build dependencies in a DAG. If `/abc` and its `zlib` are link dependencies, though, we know it'll fail to concretize.

On the other hand, we currently do this:

```console
> spack spec tcl ^zlib@1.2 ^zlib@1.3
==> Error: Cannot depend on incompatible specs 'zlib@1.2' and 'zlib@1.3'
```

IMO this is a useful check for users, since it's most likely a mistake, and if (as we've said) our intent is to make `^` mean "unified spec", i.e., a transitive link or run dependency of the root, not some dependency of a build dep, then it's reporting a real error early. I guess `tcl ^zlib@1.2 ^%zlib@1.3` would be ok in our future model.

The inconsistency, I guess, is that we detect this earlier at parse time than we do in `lookup_hash`.  Technically if you had a choice between two `/abc`'s, we *could* be smart enough to choose the one with the satisfiable `zlib` configuration here.  We do that already for constraints *on the hash spec*; it's just for transitive abstract hashes that we aren't considering the whole spec.  I guess this isn't so bad.  It's a corner case where we might complain about an abstract hash being ambiguous when we don't need to, but I think I could come up with cases where you'd *have* to solve to disambiguate, and I wouldn't expect a lookup to imply a solve.

Ok, I think I convinced myself that this is fine.  I just wanted to document this and get others' opinions.